src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+                                      haddr_t addrs[], size_t sizes[], const void *bufs[])
+{
+    const void **copied_bufs       = NULL;
+    H5FD_mem_t  *copied_types      = NULL;
+    haddr_t     *copied_addrs      = NULL;
+    size_t      *copied_sizes      = NULL;
+    H5FD_mem_t   type              = H5FD_MEM_DEFAULT;
+    size_t       io_size           = 0;
+    bool         all_super_writes  = true;
+    bool         some_super_writes = false;
+    bool         extend_types      = false;
+    bool         extend_sizes      = false;
+    herr_t       ret_value         = SUCCEED;
+
+    assert(file_ptr);
+
+    /* Only mirror I/O from MPI rank 0 */
+    if (file_ptr->mpi_rank != 0)
+        H5_SUBFILING_GOTO_DONE(SUCCEED);
+
+    if (count == 0)
+        H5_SUBFILING_GOTO_DONE(SUCCEED);
+
+    for (size_t i = 0; i < count; i++) {
+        if (!extend_types) {
+            if ((i > 0) && (types[i] == H5FD_MEM_NOLIST)) {
+                extend_types = true;
+                type         = types[i - 1];
+            }
+            else {
+                type = types[i];
+            }
+        }
+
+        if (type == H5FD_MEM_SUPER)
+            some_super_writes = true;
+        else
+            all_super_writes = false;
+    }
+
+    if (all_super_writes) {
+        if (H5FD_write_vector(file_ptr->stub_file, count, types, addrs, sizes, bufs) < 0)
+            H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL,
+                                    "couldn't write superblock information to stub file");
+    }
+    else if (some_super_writes) {
+        uint32_t vec_len = 0;
+
+        /* Copy I/O vectors and strip out non-superblock I/O */
+
+        if (NULL == (copied_types = malloc(count * sizeof(*copied_types))))
+            H5_SUBFILING_GOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                    "can't allocate copy of I/O types array");
+        if (NULL == (copied_addrs = malloc(count * sizeof(*copied_addrs))))
+            H5_SUBFILING_GOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                    "can't allocate copy of I/O addresses array");
+        if (NULL == (copied_sizes = malloc(count * sizeof(*copied_sizes))))
+            H5_SUBFILING_GOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                    "can't allocate copy of I/O sizes array");
+        if (NULL == (copied_bufs = malloc(count * sizeof(*copied_bufs))))
+            H5_SUBFILING_GOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
+                                    "can't allocate copy of I/O buffers array");
+
+        extend_types = false;
+        extend_sizes = false;
+        for (size_t i = 0; i < count; i++) {
+            if (!extend_types) {
+                if ((i > 0) && (types[i] == H5FD_MEM_NOLIST)) {
+                    extend_types = true;
+                    type         = types[i - 1];
```

Comment:
Seems silly to have a variable for type here when it can only be one thing. Not that it makes a huge difference

Reply:
For now, I've just left this the same since it matches other places and was easier for me to reason about.